internal/provider/providers/namesilo/provider.go

```diff
@@ -0,0 +1,305 @@
+	return newIP, nil
+}
+
+// https://www.namesilo.com/api-reference#dns/dns-list-records
+func (p *Provider) getRecord(ctx context.Context, client *http.Client, recordType string) (
+	id *string, ip netip.Addr, err error,
+) {
+	queryParams := url.Values{}
+	url := p.createRequestURL("/api/dnsListRecords", queryParams)
+
+	response, err := p.sendAPIRequest(ctx, client, url)
+	if err != nil {
+		return nil, netip.Addr{}, err
+	}
+
+	// find matching record
+	host := utils.BuildURLQueryHostname(p.owner, p.domain)
+	for _, record := range response.Reply.Records {
+		if record.Host != host || record.Type != recordType {
+			continue
+		}
+		ip, err = netip.ParseAddr(record.Value)
+		if err != nil {
+			return nil, netip.Addr{}, fmt.Errorf("parsing existing IP: %w", err)
+		}
+		return &record.ID, ip, nil
+	}
+
+	return nil, netip.Addr{}, fmt.Errorf("%w: no matching records found", errors.ErrRecordNotFound)
+}
+
+// https://www.namesilo.com/api-reference#dns/dns-add-record
+// https://www.namesilo.com/api-reference#dns/dns-update-record
+func (p *Provider) createOrUpdateRecord(
+	ctx context.Context,
+	client *http.Client,
+	recordID *string,
+	recordType string,
+	ip netip.Addr,
+) error {
+	name := p.owner
+	if name == "@" {
+		name = ""
+	}
+
+	queryParams := url.Values{
+		"rrhost":  {name},
+		"rrvalue": {ip.String()},
+	}
+	if p.ttl != nil {
+		queryParams.Set("rrttl", strconv.FormatUint(uint64(*p.ttl), 10))
+	}
+
+	var path string
+	if recordID == nil {
+		// create new record
+		path = "/api/dnsAddRecord"
+		queryParams.Set("rrtype", recordType)
+	} else {
+		// update record by id
+		path = "/api/dnsUpdateRecord"
+		queryParams.Set("rrid", *recordID)
+	}
+
+	url := p.createRequestURL(path, queryParams)
+
+	// if the operation was successful, err will be nil
+	_, err := p.sendAPIRequest(ctx, client, url)
+	return err
+}
+
+func (p *Provider) createRequestURL(path string, queryParams url.Values) string {
+	baseURL := url.URL{
+		Scheme: "https",
+		Host:   "www.namesilo.com",
+		Path:   path,
+	}
+	queryParams.Set("version", "1")
+	queryParams.Set("type", "json")
+	queryParams.Set("key", p.key)
+	queryParams.Set("domain", p.domain)
+	baseURL.RawQuery = queryParams.Encode()
+	return baseURL.String()
+}
+
+func (p *Provider) sendAPIRequest(ctx context.Context, client *http.Client, url string) (*APIResponse, error) {
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating http request: %w", err)
+	}
+	headers.SetUserAgent(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	data, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading response body: %w", err)
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.ToSingleLine(string(data)))
```

Comment:
do they return any structured error body i.e. json when the status code isn't ok?
Also, when the record is created, do they return status ok or status created (201)?

Reply:
Unfortunately, this is not documented.

From my testing, the server always responds with a HTTP OK (200). This includes error flows (like invalid key or missing mandatory parameters) and record creation. I assume it's possible for the server to return a non-200 during server outages or similar, but I do not know how the error body is structured.